### ISSUE TITLE
Question about lifetimes now that Response includes a reference to Headers

### ISSUE THREAD

**Issue by jimmycuadra** [hyperium/hyper#539]
In 0.5.0, the `Response` type got changed so that its `headers` field is now `&'a mut Headers` instead of just `Headers`. (changed in 49b5b8f)

I had some private functions in a program I'm working on that looked like this:

```
use hyper::Client;
use hyper::Result as HyperResult;
use hyper::client::Response;
use hyper::header::ContentType;
use hyper::method::Method;

// ... snip ...

fn request(method: Method, url: String) -> HyperResult<Response> {
    let mut client = Client::new();

    client.request(method, &url).send()
}

fn request_with_body(method: Method, url: String, body: String) -> HyperResult<Response> {
    let mut client = Client::new();

    client.request(method, &url).header(ContentType::json()).body(&body).send()
}

```

This now fails to compile because the hyper `Result` getting returned now has a reference in it where it didn't before:

```
src/http.rs:28:5: 28:11 error: `client` does not live long enough
src/http.rs:28     client.request(method, &url).send()
                   ^~~~~~
src/http.rs:25:66: 29:2 note: reference must be valid for the destruction scope surrounding block at 25:65...
src/http.rs:25 fn request(method: Method, url: String) -> HyperResult<Response> {
src/http.rs:26     let mut client = Client::new();
src/http.rs:27
src/http.rs:28     client.request(method, &url).send()
src/http.rs:29 }
src/http.rs:26:36: 29:2 note: ...but borrowed value is only valid for the block suffix following statement 0 at 26:35
src/http.rs:26     let mut client = Client::new();
src/http.rs:27
src/http.rs:28     client.request(method, &url).send()
src/http.rs:29 }
```

I could change the public API of my module so that you create a struct which contains the Hyper client in the calling function, but it was nice to have it work as a single function. Would this be the recommended way to update this code?

In other words, how can I encapsulate "creating a client, making a call, and returning the Hyper result" now that the client must outlast the result?

Thank you!

**Comment by jimmycuadra** [comment]
Never mind... I'm an idiot. The commit I linked to was about the server and does not affect the client. What I needed to do was create a let binding to the Request before calling `send` on it:

``` rust
fn request(method: Method, url: String) -> HyperResult<Response> {
    let mut client = Client::new();

    let request = client.request(method, &url);

    request.send()
}
```

**Comment by seanmonstar** [comment]
That looks odd to me. As you can see in the various examples in hyper's
docs, I chain the methods until I get a Response...

**Comment by jimmycuadra** [comment]
Could it be because I was attempting to return the `Result` without unwrapping it?

**Comment by seanmonstar** [comment]
Yea, something wonky with the borrow checker.

```rust
fn request(url) -> Result<Response> {
    let mut client = Client::new();
    client.get(&url).send() // borrowck
}

fn request(url) -> Result<Response> {
    let mut client = Client::new();
    let res = client.get(&url).send();
    res // um ok
}
```
